asciidoc/modules/con_mg_changes-in-key-management.adoc

```diff
@@ -0,0 +1,121 @@
+
+The following classes have been updated:
+
+* `io.vertx.ext.auth.KeyStoreOptions` used to work with `jce` keystores
+
+* `io.vertx.ext.auth.SecretOptions` used to handle symmetric secrets
+
+* `io.vertx.ext.auth.PubSecKeyOptions` used to handle public secret keys
+
+The following section describes the changes in key management.
+
+== Secret options class is no longer available
+
+The `SecretOptions` class is no longer available. Use the new `PubSecKeyOptions` class instead to work with a cryptographic key.
+
+The following example shows how methods of `SecretOptions` class were used in {VertX} {3x} releases.
+
+----
+new SecretOptions()
+    .setType("HS256")
+    .setSecret("password")
+----
+
+The following example shows how methods of `PubSecKeyOptions` class should be used in {VertX} {4}.
+
+----
+new PubSecKeyOptions()
+    .setAlgorithm("HS256")
+    .setSecretKey("password")
+----
+
+== Updates in public secret keys management
+
+In {VertX} {3x}, the configuration object in public secret key management assumed that:
+
+* Keys are configured as key-pairs.
+* Key data is a PKCS8 encoded string without standard delimiters.
+
+The following example shows how to configure key pair in {VertX} {3x}.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+new PubSecKeyOptions()
+  .setPublicKey(
+    // remove the boundaries
+    // (don't do this as it doesn't handle end of lines
+    // it's for illustration purposes)
+    pubPemString
+      .replaceAll("-----BEGIN PUBLIC KEY----")
+      .replaceAll("-----END PUBLIC KEY----"))
+  .setSecretKey(
+    // remove the boundaries
+    // (don't do this as it doesn't handle end of lines
+    // it's for illustration purposes)
+    secPemString
+      .replaceAll("-----BEGIN PUBLIC KEY----")
+      .replaceAll("-----END PUBLIC KEY----"));
+----
+
+In {VertX} {4}, the configuration uses a single key, so a key-pair should be 2 key (the public
+the secret)XXX.
+
+The following example shows how to configure key pair in {VertX} {4}.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+PubSecKeyOptions pubKey =
+  new PubSecKeyOptions()
+    // the buffer is the exact contents of the PEM file (boundaries included)
+    .setBuffer(pubPemString);
+
+PubSecKeyOptions secKey =
+  new PubSecKeyOptions()
+    // the buffer is the exact contents of the PEM file (boundaries included)
+    .setBuffer(secPemString);
+----
+
+You can now handle X509 certificates using `PubSecKeyOptions`.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+PubSecKeyOptions x509Certificate =
+  new PubSecKeyOptions()
+    // the buffer is the exact contents of the PEM file (boundaries included)
+    .setBuffer(x509PemString);
+----
+
+== Changes in keystore management
+
+In {VertX} {3x}, `KeyStoreOptions` assumes that the keystore format is `jceks`, and the stored password is the same as the password of the key. As `jceks` is a proprietary format, it is recommended to use a standard format, such as JDK, instead.
```

Comment:
@vietj What is new in Vert.x 4.0 with respect to keystore management?

Reply:
@pmlopes can you address this concern ?

Reply:
@sangeetaraghu In 4.0 `KeyStoreOptions` will now allow choosing the store type. (By default, when unspecified it will assume `jceks` for backwards compatibility, but users can now specify other types such as `PKCS11`, `PKCS12`, etc...).

A second improvement is that users can now use keystores that have dedicated passwords for each keystore entry. In 3.x it was assumed that all entries would share the same password, the keystore password. In 4, A password for each entry is allowed to be configured with fallback to the keystore password when no entry specific one is provided.

Reply:
@pmlopes I have updated the text as below. Please let me know if it reads ok:

When you use `KeyStoreOptions` in {VertX} {4}, you can specify a store type. For example, store types such as PKCS11, PKCS12, and so on can be set. The default store type is `jceks`. 

In {VertX} {3x}, all keystore entries would share the same password, that is, the keystore password. In {VertX} {4}, each keystore entry can have a dedicated password. If you do not want to set password for each keystore entry, you can configure the keystore password as the default password for all entries.